File: java/src/ch1/Ex128.java
package ch1;

import java.util.*;
import java.util.function.*;

public class Ex128 {
    public static void main(String[] args) {
        System.out.println(test_case(10));
        System.out.println(test_case(11));
        System.out.println(test_case(561));
        System.out.println(test_case(1105));
        System.out.println(test_case(1729));
        System.out.println(test_case(2465));
        System.out.println(test_case(2821));
        // This is too big???
        // System.out.println(test_case(6601));
        System.out.println(test_case(6603));
    }

    static String test_case(int n) {
        return String.format("%d \t Prime: %s \t Fast Prime: %s \t Miller-Rabin: %s", 
                             n, prime.apply((long)n), fast_prime.apply(n, 20), fast_miller_rabin.apply(n, 20));
    }

    static class Expmod {
        final long b, e, m;
        Expmod(long base, long exp, long modulus) {
            b = base; e = exp; m = modulus;
        }    

        long value() {
            if (e == 0) 
                return 1;
            else if (divides.apply(2L, e))
                return square.apply(new Expmod(b, e/2, m).value()) % m;
            else
                return (b * (new Expmod(b, e-1, m).value())) % m;
        }    
    }

    static class MillerRabinExpmod extends Expmod {
        MillerRabinExpmod(long base, long exp, long modulus) {
            super(base, exp, modulus);
        }    

        long nontrivial_square_root(long x, long sq) {
            boolean result = sq == 1L && x != 1L && x != (m-1);
            return result ? 0L : sq;
        }    

        long squaremod_with_check(long n) {
            return nontrivial_square_root(n, square.apply(n) % m);
        }    

        long value() {
            if (e == 0)
                return 1;
            else if (divides.apply(2L, e))
                return squaremod_with_check(new MillerRabinExpmod(b, e/2, m).value());
            else
                return (base * (new MillerRabinExpmod(b, e-1, m).value())) % m;
        }    
    }    

    static BiFunction<Integer, Integer, Boolean> fast_miller_rabin;
    static {
        fast_miller_rabin = (n, times) ->
        times == 0 ||
        (miller_rabin_test(n) && fast_miller_rabin.apply(n, times-1));
    }    

    static boolean miller_rabin_test(int n) {
        Function<Integer, Boolean> f = a -> 1 == new MillerRabinExpmod(a, n-1, n).value();
        return f.apply(1 + random.apply(n-1));
    }

    static BiFunction<Integer, Integer, Boolean> fast_prime;
    static {
        fast_prime = (n, times) ->
        times == 0 ||
        (fermat_test(n) && fast_prime.apply(n, times-1));
    }    

    static boolean fermat_test(int n) {
        Function<Integer, Boolean> f = a -> a == new Expmod(a, n, n).value();
        return f.apply(1 + random.apply(n-1));
    }

    static Function<Integer, Integer> random = limit -> 
        new Random().nextInt(limit);

    static Function<Long, Long> square = n -> 
        n * n;

    static BiFunction<Long, Long, Boolean> divides = (a, b) -> 
        0 == b % a;

    static BiFunction<Long, Long, Long> find_divisor;
    static {
        find_divisor = (n, test) ->
        square.apply(test) > n ? 
        n : 
        divides.apply(test, n) ? test : find_divisor.apply(n, test+1);
    }

    static Function<Long, Long> smallest_divisor = n -> 
        find_divisor.apply(n, 2L);

    static Function<Long, Boolean> prime = n ->
        n == smallest_divisor.apply(n);
}

File: clojure/src/ch1/ex128.clj
;; (load-file "filename.clj")
;; or on command line> clj filename.clj


(def FAST-TIMES 20)


(defn divides? [a b]
  (= (rem b a) 0))


(defn square [n]
  (* n n))


(defn expmod [base exp m]
  (cond (= exp 0) 1
        (not (odd? exp))
          (rem (square (expmod base (/ exp 2) m)) m)
        :else 
          (rem (* base (expmod base (dec exp) m)) m)))
  

(defn miller-rabin-expmod [base exp m]
    (let [nontrivial-sqrt (fn [x sq]
                              (if (and (= sq 1)
                                       (not (= x 1))
                                       (not (= x (dec m))))
                                  0
                                  sq))
          squaremod-with-check (fn [x]
                                   (nontrivial-sqrt x (rem (square x) m)))]
        (cond (= exp 0) 1
              (even? exp) (squaremod-with-check (recur base (/ exp 2) m))
              :else (rem (* base (miller-rabin-expmod base (dec exp) m)) m))))


(defn fermat-test [n]
  (let [try-it (fn [a] (= (expmod a n n) a))]
    (try-it (inc (rand-int (dec n))))))


(defn fast-prime? [n times]
  (cond (= times 0) true
        (fermat-test n) (fast-prime? n (dec times))
        :else false))


(defn find-divisor [n test-divisor]
    (cond (> (square test-divisor) n) n
          (divides? test-divisor n) test-divisor
          :else (find-divisor n (inc test-divisor))))


(defn smallest-divisor [n]
    (find-divisor n 2))
    

(defn prime? [n]
  (= n (smallest-divisor n)))
  
  
(defn miller-rabin-test [n]
  (let [try-it (fn [a] (= (miller-rabin-expmod a (dec n) n) 1))]
    (try-it (inc (rand-int (dec n))))))


(defn fast-miller-rabin? [n times]
    (or (= times 0)
        (and (miller-rabin-test n)
             (recur n (dec times)))))


(defn test-case [n]
  (let [p format]
    (p "%d  \tPrime: %s \tFast Prime: %s \tMiller-Rabin: %s" 
    n (prime? n) (fast-prime? n FAST-TIMES) (fast-miller-rabin? n FAST-TIMES))))


(require
 'clojure.test)

(println (test-case 10))
(println (test-case 11))
(println (test-case 13))
(println (test-case 561))
(println (test-case 1105))
(println (test-case 1729))
(println (test-case 2465))
(println (test-case 6601))
(println (test-case 6603))

(println "done")